com/williamfiset/algorithms/dp/VirtualMachinesScheduler.java

```diff
@@ -0,0 +1,177 @@
+import java.util.Collections;
+import java.util.Scanner;
+
+/**
+ * Cloud computing companies, like Amazon, have a lot of virtual machines in their disposal, which
+ * they offer to users for a price for the total time of the VM usage. There are quite a few
+ * different types of VMs with different CPU and Disk speeds as well as storage sizes, which all
+ * have different costs. For example VM-A has faster CPUs than VM-B, but VM-B has faster storage
+ * read/write. Then, there is VM-C which is faster than both VM-A and VM-B in both CPU and Disk
+ * speeds but is considerably more expensive. The problem that arises is how to assign the processes
+ * to the VMs efficiently and we achieve that with dynamic programming.
+ *
+ * <p>We have a complex process that consists of N steps in a chain: P1 -> P2 -> P3 -> ... -> Pn We
+ * also have M different VMs.
+ *
+ * <p>As an input we have 2 arrays, one being NxM showing the cost of a process running in a VM, and
+ * the other being MxM showing the cost of sending data from one VM to an other.
+ *
+ * <p>E.g
+ *
+ * <p>4 Processes 3 Virtual Machines
+ *
+ * <p>5 6 3 Each process has 4 steps and there are 3 VMs. 7 8 5 The first VM runs the 4 processes
+ * with 5, 7, 7, 2 cost respectively and so on. 7 8 3 2 7 6
+ *
+ * <p>0 7 2 7 0 2 2 2 0
+ *
+ * <p>The Algorithm fills an array Costs NxM, where each cell Cost(i, j) contains the minimum total
+ * cost of the complex process until step i when it runs on VM-j. In order for step i to run on
+ * VM-j, the results from i-1 step must already be on VM-j or sent on VM-j including the
+ * communication cost between VMs.
+ *
+ * <p>For the example above the Cost array is the following:
+ *
+ * <p>5 6 3 12 13 8 17 18 11 15 20 17
+ *
+ * <p>This means that the least expensive execution costs 15 and is achieved when the last process
+ * is on the first VM.
+ *
+ * @author Dragatis Nikolas, nikosdraga@gmail.com
+ */
+public class VirtualMachinesScheduler {
+  public static void main(String[] args) throws Exception {
+    double time = System.currentTimeMillis();
+    Scheduler scheduler = new Scheduler();
+
+    scheduler.scanFile(args[0]);
+
+    System.out.println("\nTime: " + (System.currentTimeMillis() - time) / 1000.0 + " seconds.");
+  }
+
+  /**
+   * This is the class that contains the methods for producing the final array of the dynamic
+   * programming after reading and storing the data from a txt file.
+   */
+  public static class Scheduler {
+    private int virtualMachines = 0;
+    private int processes = 0;
+
+    /**
+     * This function is used to scan the txt file, with the data, and initialize the dimensions of
+     * the array with the processes costs and the array with the communication costs.
+     *
+     * @param pathname is the name of the txt file containing the data.
+     * @throws Exception
+     */
+    public void scanFile(String pathname) throws Exception {
+      Scanner scanner = new Scanner(new FileReader(pathname));
```

Comment:
Wouldn't it be more practical to pass in the number of processes and VMs as arguments to this method and avoid the need to have a scanner at all? This way the user doesn't even need a file, they can get the data from anywhere.

Reply:
For me it's easier to use a file because you can easily modify it quickly and save it and use it on the go. I could change it though, if it's a problem.

Reply:
So you mean to pass this data as arguments to the program, right?

Reply:
I changed the way the input is given as you suggested. It's now given as arguments to the program.